Add DPDK Support

tbarbette opened the pull request:
Provide 3 new elements to use Intel's DPDK framework (compatible from
version 1.7 to the current 2.1) to receive and output packets at very
high speed in Click. Configure using the --enable-dpdk to enable the
DPDK framework. [1] provides a comparison of throughput using the new
DPDK implementation and the Netmap one with the best configuration
achievable in a forwarding (without touching the packet) and a router
configuration.

FromDpdkDevice is used to receive packets and ToDpdkDevice to output
packets. DpdkInfo can be used to set some DPDK-related parameters, see
[2] and dpdkinfo.hh for more informations about them.

These I/O elements supports multiqueue, you can use the QUEUE
parameter to specify a queue index. If omitted, one queue will be used
per FromDpdkDevice/ToDpdkDevice

Difference from traditional ToDevice is that ToDpdkDevice only
supports PUSH mode, as DPDK does not allow select-like mechanism.
Anyway [3] shows that full-push mode leads to better performance most
of the time. However, under very low throughput this could induce
higher latency as the output device waits for a certain number of
packets (BURST parameter) before flushing. The TIMEOUT parameter
allows you to limit that time.

ToDpdkDevice is thread-safe, by using a vector of internal queues to
batch packets, and one timer per thread. It uses a spinlock when the
batch is ready to protect the physical device. Regarding the
FromDpdkDevice, as only one thread will spawn the Task, it does not
need such a lock or per-thread state. This mechanism is by far faster
than using a ThreadSafeQueue or an external locking element.

This patch is a stripped version of the DPDK support in FastClick
(mostly without batching and auto-thread assignment). User may want to
read [3] for implementation detail, better queue/burst usage and read
further about full-push mode.

For archive, we also tried a DPDK-pool mode, where we don't use Click
Packet object but a wrapper around DPDK's mbuf. While this saves a
little memory, it was slower as Click's Packet object are re-used and
provide a better memory locality.
Indeed, putting anotations in DPDK's mbuf header lead to two cache
miss per header instead of zero when Click can cycle through Packet
object, just changing the buffer pointers and reusing them. If people
are interested in this, they can try it in the code of [3].

[1] http://bit.ly/1Z2Eg4W - Performance of this new DPDK implementation
[2] http://www.dpdk.org - DPDK website
[3] http://hdl.handle.net/2268/181954 - FastClick paper

bcronje commented:
Great stuff @tbarbette 
I for one hope this lands in mainline Click.

kohler commented on INSTALL:
```diff
@@ -328,6 +328,42 @@ have a look at the following Xen toolstack:
+environment variable must be set, e.g.
+
+    export RTE_SDK=/path/to/dpdk-2.1.0/
+    export RTE_TARGET=x86_64-native-linuxapp-gcc
```
I think you should suggest instead

```
./configure RTE_SDK=... RTE_TARGET=... --enable-dpdk --enable-user-multithread
```

since then ./config.status will remember the settings.

kohler commented on INSTALL:
```diff
@@ -328,6 +328,42 @@ have a look at the following Xen toolstack:
+    -n NUM      : number of memory channels
+
+Note that Click's `-j` (or `--threads`) argument is not used in DPDK mode. The
+number of threads is set indirectly through the `-c` argument.
```
A little warty...

kohler commented on elements/userlevel/dpdkinfo.cc:
```diff
@@ -0,0 +1,52 @@
+    }
+    instance = this;
+    if (Args(conf, this, errh)
+        .read_p("NB_MBUF", DpdkDevice::NB_MBUF)
```
These should all just be `read` rather than `read_p`, since they aren't positional arguments.

I would prefer DPDKInfo, and similar namings; Click generally uses all-caps for acronyms (e.g. CheckTCPHeader).

kohler commented on elements/userlevel/fromdpdkdevice.cc:
```diff
@@ -0,0 +1,119 @@
+
+    if (Args(conf, this, errh)
+        .read_mp("PORT_ID", _port_id)
+        .read_p("QUEUE_ID", _queue_id)
```
Should be `QUEUE` based on documentation

kohler commented on elements/userlevel/todpdkdevice.hh:
```diff
@@ -0,0 +1,174 @@
+
+=c
+
+ToDpdkDevice(PORT [, QUEUE, IQUEUE, BLOCKING, BURST, TIMEOUT, NDESC]])
```
Oops! The configuration argument order defined in the source isn't in sync (you have BLOCKING, IQUEUE in the source). Suggest you make these keyword arguments only; the documentation would be updated with `I<keywords>` and the `read_p`s would change to `read`.

kohler commented on elements/userlevel/dpdkinfo.cc:
```diff
@@ -0,0 +1,52 @@
+        .read_p("RX_WTHRESH", DpdkDevice::RX_WTHRESH)
+        .read_p("TX_PTHRESH", DpdkDevice::TX_PTHRESH)
+        .read_p("TX_HTHRESH", DpdkDevice::TX_HTHRESH)
+        .read_p("TX_WTHRESH", DpdkDevice::TX_WTHRESH)
```
I think it would be best to rethink this design a bit—we should store these parameters once per Router, rather than globally, so that (for instance) RX_PTHRESH setting would be reset after a live reconfiguration. But that is a nit.

kohler commented on include/click/algorithm.hh:
```diff
@@ -90,5 +90,8 @@ struct less {
 
+#define click_min(a, b) ((a) < (b) ? (a) : (b))
+#define click_max(a, b) ((a) > (b) ? (a) : (b))
+
```
:cry:
Must we? You only use click_min once and you use click_max not at all. Let's scotch this.

kohler commented on userlevel/click.cc:
```diff
@@ -609,12 +632,20 @@ main(int argc, char **argv)
       }
+#else
+# if HAVE_DPDK
+      errh->warning("Click was built with DPDK support, use EAL core mask to set threads");
```
It would be better to handle this differently but it's not a big deal.

kohler commented:
Tom, this is beautiful. I left some nit-type line notes. But if you push back we can just merge right away. THANK YOU!!!!

tbarbette commented:
You're welcome! We'll look at your comments and push some changes by tomorrow ;)

tbarbette commented:
The problem of the threads handled by DPDK is with all DPDK apps. DPDK EAL arguments are to passed before a "--" and the app/Click argument after. It's somehow complicated to change how parameters are handled, DPDK is not just a normal library. But we updated the doc to explain it better. 

Regarding the DPDKInfo globality, where would you store them (in the Router class?), and how would you set them? Anyway this could be changed in a second time.

Apart from that, I think we corrected all other issues ;)

kohler commented:
*Merged With Joy*

bcronje commented:
Awesome, thank you Tom and Eddie!
